cli/src/command.rs

```diff
@@ -150,33 +150,172 @@ pub fn run() -> Result<()> {
+						service::kusama_runtime::RuntimeApi,
+						service::KusamaExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.database), task_manager))
+				})
+			} else if chain_spec.is_westend() {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::westend_runtime::RuntimeApi,
+						service::WestendExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.database), task_manager))
+				})
+			} else {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::polkadot_runtime::RuntimeApi,
+						service::PolkadotExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.database), task_manager))
+				})
+			}
+		},
+		Some(Subcommand::ExportState(cmd)) => {
+			let runner = cli.create_runner(cmd)?;
+			let chain_spec = &runner.config().chain_spec;
+
+			set_default_ss58_version(chain_spec);
+
+			if chain_spec.is_kusama() {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::kusama_runtime::RuntimeApi,
+						service::KusamaExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.chain_spec), task_manager))
+				})
+			} else if chain_spec.is_westend() {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::westend_runtime::RuntimeApi,
+						service::WestendExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.chain_spec), task_manager))
+				})
+			} else {
+				runner.async_run(|mut config| {
+					let (client, _, _, task_manager) = service::new_chain_ops::<
+						service::polkadot_runtime::RuntimeApi,
+						service::PolkadotExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, config.chain_spec), task_manager))
+				})
+			}
+		},
+		Some(Subcommand::ImportBlocks(cmd)) => {
+			let runner = cli.create_runner(cmd)?;
+			let chain_spec = &runner.config().chain_spec;
+
+			set_default_ss58_version(chain_spec);
+
+			if chain_spec.is_kusama() {
+				runner.async_run(|mut config| {
+					let (client, _, import_queue, task_manager) = service::new_chain_ops::<
+						service::kusama_runtime::RuntimeApi,
+						service::KusamaExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, import_queue), task_manager))
+				})
+			} else if chain_spec.is_westend() {
+				runner.async_run(|mut config| {
+					let (client, _, import_queue, task_manager) = service::new_chain_ops::<
+						service::westend_runtime::RuntimeApi,
+						service::WestendExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, import_queue), task_manager))
+				})
+			} else {
+				runner.async_run(|mut config| {
+					let (client, _, import_queue, task_manager) = service::new_chain_ops::<
+						service::polkadot_runtime::RuntimeApi,
+						service::PolkadotExecutor,
+					>(&mut config)?;
+					Ok((cmd.run(client, import_queue), task_manager))
+				})
+			}
+		},
```

Comment:
Would it be possible for us to have an fn that does all this?

and we call it so

`async_run(&cli, cmd)`

where `cmd` is a generic trait with the `run` method

and it internally does all these runtime checks?

Reply:
So I tried, this and came up with:
```rust

fn run_cmd<CMD, FN, B>(cli: &Cli, cmd: &CMD, func: FN) -> Result<()>
	where
		FN: FnOnce(sc_cli::Runner<Cli>, &CMD) -> Result<()>,
		CMD: sc_cli::CliConfiguration + sc_cli::SubstrateCli,
{
	let runner = cli.create_runner(cmd)?;
	let chain_spec = &runner.config().chain_spec;

	set_default_ss58_version(chain_spec);

	if chain_spec.is_kusama() {
		func::<service::kusama_runtime::Block, service::KusamaExecutor>(runner, cmd)
	} else if chain_spec.is_westend() {
		func::<service::westend_runtime::Block, service::WestendExecutor>(runner, cmd)
	} else {
		func::<service::polkadot_runtime::Block, service::PolkadotExecutor>(runner, cmd)
	}
}
```

Unfortunately, rust does not allow you to put generics on a `FnOnce` like this:

```
error[E0109]: type arguments are not allowed for this type
   --> cli/src/command.rs:366:10
    |
366 |         func::<service::kusama_runtime::Block, service::KusamaExecutor>(runner, cmd)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type argument not allowed

error[E0109]: type arguments are not allowed for this type
   --> cli/src/command.rs:368:10
    |
368 |         func::<service::westend_runtime::Block, service::WestendExecutor>(runner, cmd)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type argument not allowed

error[E0109]: type arguments are not allowed for this type
   --> cli/src/command.rs:370:10
    |
370 |         func::<service::polkadot_runtime::Block, service::PolkadotExecutor>(runner, cmd)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type argument not allowed

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0109`.
error: could not compile `polkadot-cli`.

```

You get the same problem with a macro as well.

Reply:
What about: 

```rust

fn run_cmd<CMD, B>(cli: &Cli, cmd: &CMD) -> Result<()>
	where
		CMD: sc_cli::CliConfiguration + sc_cli::SubstrateCli,
{
	let runner = cli.create_runner(cmd)?;
	let chain_spec = &runner.config().chain_spec;

	set_default_ss58_version(chain_spec);

	if chain_spec.is_kusama() {
    runner.async_run(|mut config| {
		  let (client, _, import_queue, task_manager) = service::new_chain_ops::<
			  service::kusama_runtime::RuntimeApi,
			  service::KusamaExecutor,
		  >(&mut config)?;
		  Ok((cmd.run(client, import_queue), task_manager))
	  })
	} else if chain_spec.is_westend() {
    runner.async_run(|mut config| {
		  let (client, _, import_queue, task_manager) = service::new_chain_ops::<
			  service::westend_runtime::RuntimeApi,
			  service::WestEndExecutor,
		  >(&mut config)?;
		  Ok((cmd.run(client, import_queue), task_manager))
	  })
	} else {
    runner.async_run(|mut config| {
		  let (client, _, import_queue, task_manager) = service::new_chain_ops::<
			  service::polkadot_runtime::RuntimeApi,
			  service::PolkadotExecutor,
		  >(&mut config)?;
		  Ok((cmd.run(client, import_queue), task_manager))
	  })
	}
}

```

Reply:
The different subcommands have different inputs unfortunately.

Reply:
In that case i think it's fine for now

Reply:
I think you can achieve this if you use the Client enum that Basti introduced recently: https://github.com/paritytech/polkadot/blob/master/service/src/client.rs#L138

You would need to make the cmd.run accepts that Client enum.

Ideally we should move the `if kusama { ... } else if westend { ... } else ... { ... }` pattern to polkadot service instead of the binary.

Reply:
Yep, that works.